File: 品味Java的21種設計模式/dp_cpp/prototype/example6/PersonalOrder.cpp
#include "PersonalOrder.h"

namespace cn {
namespace javass {
namespace dp {
namespace prototype {
namespace example6 {

int PersonalOrder::getOrderProductNum() {
    return this->orderProductNum;
}

void PersonalOrder::setOrderProductNum(int num) {
    this->orderProductNum = num;
}

std::string PersonalOrder::getCustomerName() {
    return customerName;
}

void PersonalOrder::setCustomerName(std::string customerName) {
    this->customerName = customerName;
}

Product *PersonalOrder::getProduct() {
    return product;
}

void PersonalOrder::setProduct(Product *product) {
    this->product = product;
}

std::string PersonalOrder::ToString() {
    //简单点输出
    return "订购产品是=" + this->product->getName() + "，订购数量为=" + this->orderProductNum;
}

OrderApi *PersonalOrder::cloneOrder() {
    //创建一个新的订单，然后把本实例的数据复制过去
    PersonalOrder *order = new PersonalOrder();
    order->setCustomerName(this->customerName);
    order->setOrderProductNum(this->orderProductNum);
    //对于对象类型的数据，深度克隆的时候需要继续调用这个对象的克隆方法
    order->setProduct(static_cast<Product *>(this->product->cloneProduct()));
    return order;
}
}
}
}
}
}

File: 品味Java的21種設計模式/dp_java/prototype/example6/PersonalOrder.java
package cn.javass.dp.prototype.example6;
/**
 * 个人订单对象
 */
public class PersonalOrder implements OrderApi {
    /**
     * 订购人员姓名
     */
    private String customerName;
    /**
     * 产品对象
     */
    private Product product = null;
    /**
     * 订单产品数量
     */
    private int orderProductNum = 0;

    public int getOrderProductNum() {
        return this.orderProductNum;
    }
    public void setOrderProductNum(int num) {
        this.orderProductNum = num;
    }
    public String getCustomerName() {
        return customerName;
    }
    public void setCustomerName(String customerName) {
        this.customerName = customerName;
    }
    public Product getProduct() {
        return product;
    }
    public void setProduct(Product product) {
        this.product = product;
    }

    public String toString() {
        //简单点输出
        return "订购产品是=" + this.product.getName() + "，订购数量为=" + this.orderProductNum;
    }

    public OrderApi cloneOrder() {
        //创建一个新的订单，然后把本实例的数据复制过去
        PersonalOrder order = new PersonalOrder();
        order.setCustomerName(this.customerName);
        order.setOrderProductNum(this.orderProductNum);
        //对于对象类型的数据，深度克隆的时候需要继续调用这个对象的克隆方法
        order.setProduct((Product)this.product.cloneProduct());
        return order;
    }
}
